scalding-core/src/main/scala/com/twitter/scalding/typed/CascadingBackend.scala

```diff
@@ -0,0 +1,268 @@
+import java.util.WeakHashMap
+
+object CascadingBackend {
+  import TypedPipe._
+
+  /**
+   * we want to cache renderings of some TypedPipe to Pipe so cascading
+   * will see them as the same. Without this, it is very easy to have
+   * a lot of recomputation. Ideally we would plan an entire graph
+   * at once, and not need a static cache here, but currently we still
+   * plan one TypedPipe at a time.
+   */
+  private[this] val pipeCache = new WeakHashMap[TypedPipe[Any], Map[Mode, CascadingPipe[Any]]]()
+
+  private def cacheGet[T](t: TypedPipe[T], m: Mode)(p: FlowDef => CascadingPipe[T]): CascadingPipe[T] = {
+    def add(mmc: Map[Mode, CascadingPipe[Any]]): CascadingPipe[T] = {
+      val emptyFD = new FlowDef
+      val res = p(emptyFD)
+      pipeCache.put(t, mmc + (m -> res.asInstanceOf[CascadingPipe[Any]]))
+      res
+    }
+
+    pipeCache.synchronized {
+      pipeCache.get(t) match {
+        case null => add(Map.empty)
+        case somemap if somemap.contains(m) => somemap(m).asInstanceOf[CascadingPipe[T]]
+        case missing => add(missing)
+      }
+    }
+  }
+
+  final def toPipe[U](p: TypedPipe[U], fieldNames: Fields)(implicit flowDef: FlowDef, mode: Mode, setter: TupleSetter[U]): Pipe = {
+
+    def kvfields = Grouped.kvFields
+    val f0 = new Fields(java.lang.Integer.valueOf(0))
+
+    def singlePipe[T](t: TypedPipe[T], force: Boolean = false): CascadingPipe[T] =
+      cacheGet(t, mode) { localFD =>
+        val pipe = t.toPipe(f0)(localFD, mode, singleSetter)
+        val p = if (force) RichPipe(pipe).forceToDisk else pipe
+        CascadingPipe[T](p, f0, localFD, mode, singleConverter)
+      }
+
+    def loop[T](t: TypedPipe[T], rest: FlatMappedFn[T, U]): Pipe = t match {
+      case cp@CascadingPipe(_, fields, localFlowDef, m, _) =>
+        import Dsl.flowDefToRichFlowDef
+        // This check is not likely to fail unless someone does something really strange.
+        // for historical reasons, it is not checked by the typed system
+        require(m == mode,
+          s"Cannot switch Mode between TypedSource.read and toPipe calls. Pipe: $p, pipe mode: $m, outer mode: $mode")
+        flowDef.mergeFrom(localFlowDef)
+
+        def go[T1 <: T](cp: CascadingPipe[T1]): Pipe = {
+          RichPipe(cp.pipe).flatMapTo[T1, U](fields -> fieldNames)(rest)(cp.converter, setter)
+        }
+        go(cp)
+
+      case cp@CrossPipe(_, _) => loop(cp.viaHashJoin, rest)
+
+      case cv@CrossValue(_, _) => loop(cv.viaHashJoin, rest)
+
+      case DebugPipe(p) =>
+        // There is really little that can be done here but println
+        loop(p.map { t => println(t); t }, rest)
+
+      case EmptyTypedPipe =>
+        // just use an empty iterable pipe.
+        // Note, rest is irrelevant
+        IterableSource(Iterable.empty, fieldNames)(setter, singleConverter[U]).read(flowDef, mode)
+
+      case fk@FilterKeys(_, _) =>
+        def go[K, V](node: FilterKeys[K, V]): Pipe = node match {
```

Comment:
I don't want to do that because I think it actually obscures things. If scala were smarter with type inference this would not even be needed (note we just define a method to call it). I think keeping a standard pattern here actually serves to tell the reader this is doing the logic for the match, except we can name the types.